cf-testd/cf-testd.c

```diff
@@ -607,16 +643,158 @@ int main(int argc, char *argv[])
         Log(LOG_LEVEL_DEBUG, "Got report file contents: %s", config->report);
     }
 
-    Log(LOG_LEVEL_INFO, "Starting server...");
+    Log(LOG_LEVEL_INFO, "Starting server at %s...", config->address);
     fflush(stdout); // for debugging startup
 
-    int r = CFTestD_StartServer(config);
-    CFTestD_ConfigDestroy(config);
+    config->ret = CFTestD_StartServer(config);
 
     /* we don't really need to do this here because the process is about the
      * terminate, but it's a good way the cleanup actually works and doesn't
      * cause a segfault or something */
     ServerTLSDeInitialize(&(config->priv_key), &(config->pub_key), &(config->ssl_ctx));
 
-    return r;
+    LoggingPrivSetContext(prior);
+
+    return NULL;
+}
+
+static void HandleSignal(int signum)
+{
+    switch (signum)
+    {
+    case SIGTERM:
+    case SIGINT:
+        // flush all logging before process ends.
+        fflush(stdout);
+        fprintf(stderr, "Terminating...\n");
+        TERMINATE = true;
```

Comment:
@vpodzime Looks like this broke our reporting tests, because it can now take up to ~10 seconds to terminate cf-testd. Do you have a better way to do this, at least for the single threaded mode?

Reply:
See https://github.com/cfengine/core/pull/3109 and https://github.com/cfengine/nova/pull/1180